### PR TITLE
feature: investors contract

### PR DESCRIPTION
**Pull request by farrellh1**
# Description

Each company/team will have their own tokens, and the company's founder can set the token name and symbol (e.g., name: BITCOIN, symbol: BTC).

The investor contract will store a list of investor addresses and a `MintAgreement` for each investor. A `MintAgreement` consists of:

- `address investor`: the investor's address.
- `uint256 amount`: the amount the investor will receive each time the mint function is called.
- `uint256 lastMinted`: the last timestamp when the investor received tokens.
- `uint256 totalMinted`: the total tokens minted for this investor.
- `bool isActive`: the state of the agreement.

As the contract owner (founder or BOD), they can call the mint function to create tokens for all investors based on the `MintAgreement`. For example:

### Agreements:
- Founder agreement: 100 tokens
- Employee1: 20 tokens
- Employee2: 10 tokens

When the founder calls the mint function, it will mint new tokens:
- 100 tokens for the founder
- 20 tokens for Employee1
- 10 tokens for Employee2

Additionally, anyone (for now) can call the `distributeDividends` function to distribute dividends. The founder must specify how many ETH/POL will be distributed to the team. For example:

The contract owner calls the function `distributeDividends` with 100 ETH/POL.

### Agreements:
- Founder agreement: 50 tokens (50% shares)
- Employee1: 30 tokens (30% shares)
- Employee2: 20 tokens (20% shares)

### Distribution:
- 50 ETH/POL for the founder
- 30 ETH/POL for Employee1
- 20 ETH/POL for Employee2

As the contract owner, they can also:
- Add a `MintAgreement` for new investors.
- Update an investor's `MintAgreement`, where the data that can be changed includes the amount and the `isActive` status.
- Remove an investor's `MintAgreement`.
- Stop minting by calling the `pause` function.

In the future, the founder or BOD could have the authority to create an action that allows the desired amount of money to be withdrawn from a bank account, instead of allowing anyone to call the dividend distribution function.

Fixes # (issue)

## Contribution

Investor.sol -> investor contract
MintAgreement.sol -> mint agreement struct
Officer.sol -> add investor contract to officer contract

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)

## Checklist

Make sure you run the following commands before submitting your PR:

- /app

```bash
npm run build
npm run test:unit
npm run type-check
npm run lint
npm run format
```

- /backend

```bash
npm run build
npm run test
npm run lint
npm run format
```

- /contract

```bash
npm run test
npm run lint
npm run format
```

- **Please check if the PR fulfills these requirements**

- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes/features)

### CODE REVIEW

File: contract/contracts/Investor/Investor.sol [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+    function addMintAgreement(address _investor, uint256 _amount, bool _isActive) external onlyOwner whenNotPaused {
+        require(_investor != address(0), "Invalid investor address");
+        require(!investors.contains(_investor), "Investor already exists");
+        require(_amount > 0, "Invalid amount");
+
+        MintAgreement storage agreement = mintAgreements[_investor];
+        agreement.investor = _investor;
+        agreement.amount = _amount;
+        agreement.isActive = _isActive;
+
```
Comment (stravid87) [inline contract/contracts/Investor/Investor.sol:78]: My only request is here. I would like to be able to initiate a user with an amount of SHERS. Something like what you might consider a "signing bonus".

**Review by stravid87 — COMMENTED** [review]
Nice. I'm literally planning how to use this already!

**Comment by farrellh1** [comment]
I just added the signing bonus usecase @stravid87

**Review by dasarathg68 — APPROVED** [review]
LGTM!
Learned a lot of new things just by reviewing this PR. Great work!

File: contract/contracts/Investor/types/StockGrant.sol
Comment (hermannleboss) [inline contract/contracts/Investor/types/StockGrant.sol:1]: If this is not used in another file there is no need to have it here. It's complicate the experience. You need to go in a different file to know what is the data structure of StockGrant

Comment (farrellh1) [reply contract/contracts/Investor/types/StockGrant.sol:1]: it used on Officer contract

**Review by hermannleboss — COMMENTED** [review]
LGTM. 
Add the investor V1 to the officer contract. And let's think to the interface.

File: contract/contracts/Investor/InvestorV1.sol [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    __Pausable_init();
+  }
+
+  function distributeMint(address[] memory shareholders, uint256 amount) external onlyOwner whenNotPaused nonReentrant {
+    for (uint256 i = 0; i < shareholders.length; i++) {
+      address shareholder = shareholders[i];
+      _mint(shareholder, amount);
+      emit Minted(shareholder, amount);
+    }
+  }
```
Comment (hermannleboss) [inline contract/contracts/Investor/InvestorV1.sol:25]: In this case, this fuction is to distribute SHER to many user

Farrel 10
Hermann 9
Dasarath 20

**Review by hermannleboss — COMMENTED** [review]
LGTM

**Pull request merged**